File: Sorting/heapSort.py
class Heap:
    def heapify(self, arr, n, i):
        largest = i
        leftChild = 2*i + 1
        rightChild = 2*i + 2
        if leftChild < n and arr[leftChild] > arr[i]:
            largest = leftChild
        if rightChild < n and arr[rightChild] > arr[largest]:
            largest = rightChild
        if largest != i:
            arr[largest], arr[i] = arr[i], arr[largest]
            self.heapify(arr,n,largest)
    
    def buildHeap(self, arr, n):
        for i in range(n//2,-1,-1):
            self.heapify(arr,n,i)
        print(arr)
    
    def heapSort(self, arr):
        n = len(arr)
        self.buildHeap(arr,n)
        for i in range(n-1,0,-1):
            arr[0], arr[i] = arr[i], arr[0]
            self.heapify(arr,i,0)

if __name__ == '__main__':
    h = Heap()
    arr = [4,1,3,2,16,9,10,14,8,7]
    # h.buildHeap(arr, len(arr))
    print(arr)
    h.heapSort(arr)
    print(arr)

        

File: Algorithms/heapsort.cpp
#include<conio.h>
#include<iostream>
using namespace std;
void printArray(int* p,int n){
	for (int i = 0; i<n; i++){
		cout<<p[i]<<ends;
	}
	cout<<endl;
}
void swap(int * a, int *b){
	int temp;
	temp = *a;
	*b = *a;
	*a = temp;
}
void heapify(int arr[],int n,int i){
	int largest,l,r;
	l = 2*i + 1;
	r = 2*i + 2;
	if(l < n  && arr[l]>arr[i]){
		largest = l;
	}
	else
	largest = i;
	if(r < n && arr[r]>arr[i]){
		largest = r;
	}
	else
	largest = i;
	if(largest != i){
		swap(arr[largest],arr[i]);
		heapify(arr,n,largest);
	}
	
}
void built_heap(int arr[],int n){
	for (int i = n/2; i>=0; i--){
		heapify(arr,n,i);
		
	}
	printArray(arr,n);
}
void heap_sort(int arr[],int n){
	built_heap(arr,n);
	for(int i = n; i>0; i--){
		swap(arr[0],arr[i]);
		heapify(arr,i,0);
	}
}

int main(){
	int arr[] = {12,11,13,5,6,7};
	int n = sizeof(arr)/sizeof(arr[0]);
	
	heap_sort(arr,n);
	printArray(arr,n);
	
}
